unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -0,0 +1,618 @@
+     */
+    public int getValueLength() {
+      assert (isDefined);
+      return valueLength;
+    }
+
+    /**
+     * Store a new key and value. This method may only be called once for a given key; if you want
+     * to update the value associated with a key, then you can directly manipulate the bytes stored
+     * at the value address.
+     * <p>
+     * It is only valid to call this method immediately after calling `lookup()` using the same key.
+     * <p>
+     * After calling this method, calls to `get[Key|Value]Address()` and `get[Key|Value]Length`
+     * will return information on the data stored by this `putNewKey` call.
+     * <p>
+     * As an example usage, here's the proper way to store a new key:
+     * <p>
+     * <pre>
+     *   Location loc = map.lookup(keyBaseOffset, keyBaseObject, keyLengthInBytes);
+     *   if (!loc.isDefined()) {
+     *     loc.putNewKey(keyBaseOffset, keyBaseObject, keyLengthInBytes, ...)
+     *   }
+     * </pre>
+     * <p>
+     * Unspecified behavior if the key is not defined.
+     */
+    public void putNewKey(
+        Object keyBaseObject,
+        long keyBaseOffset,
+        int keyLengthBytes,
+        Object valueBaseObject,
+        long valueBaseOffset,
+        int valueLengthBytes) {
+      assert (!isDefined) : "Can only set value once for a key";
+      isDefined = true;
+      assert (keyLengthBytes % 8 == 0);
+      assert (valueLengthBytes % 8 == 0);
+      // Here, we'll copy the data into our data pages. Because we only store a relative offset from
+      // the key address instead of storing the absolute address of the value, the key and value
+      // must be stored in the same memory page.
+      final long requiredSize = 8 + 8 + keyLengthBytes + valueLengthBytes;
+      assert(requiredSize <= PAGE_SIZE_BYTES);
+      size++;
+      bitset.set(pos);
+
+      // If there's not enough space in the current page, allocate a new page:
+      if (currentDataPage == null || PAGE_SIZE_BYTES - pageCursor < requiredSize) {
+        MemoryBlock newPage = allocator.allocate(PAGE_SIZE_BYTES);
+        dataPages.add(newPage);
+        pageCursor = 0;
+        currentPageNumber++;
+        pageTable[currentPageNumber] = newPage.getBaseObject();
+        currentDataPage = newPage;
+      }
+
+      // Compute all of our offsets up-front:
+      final Object pageBaseObject = currentDataPage.getBaseObject();
+      final long pageBaseOffset = currentDataPage.getBaseOffset();
+      final long keySizeOffsetInPage = pageBaseOffset + pageCursor;
+      pageCursor += 8;
+      final long keyDataOffsetInPage = pageBaseOffset + pageCursor;
+      pageCursor += keyLengthBytes;
+      final long valueSizeOffsetInPage = pageBaseOffset + pageCursor;
+      pageCursor += 8;
+      final long valueDataOffsetInPage = pageBaseOffset + pageCursor;
+      pageCursor += valueLengthBytes;
+      final long relativeOffsetFromKeyToValue = valueSizeOffsetInPage - keySizeOffsetInPage;
```

Comment:
I think that this is redundant, actually, since the value always appears after the key, we know the key's length, and we always store the key and value in the same memory block.

If I remove this, we'll have an extra 32 bits of space in the long array that we can use for something else.  Moving to a 64-bit hashcode probably won't make a huge difference.  Is there another good use for this space?